### PR TITLE
Ast refactor

### PR DESCRIPTION
**Pull request by tudorcebere**
## Description
Second take on the ast refactor.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: src/syft/ast/globals.py [hunk 1]
```diff
@@ -7,31 +8,24 @@
...
-        return_callable: bool = False,
-        obj_type: Optional[type] = None,
     ) -> Optional[Union[Callable, CallableT]]:
-        if isinstance(path, str):
-            path = path.split(".")
-        return self.attrs[path[index]](
-            path=path,
+
+        _path: List[str] = (
+            path.split(".") if isinstance(path, str) else path if path else []
```
Comment (gmuraru) [inline src/syft/ast/globals.py:23]: Q: Shouldn't we raise an error if ```path``` is not available? (here we use an empty list, but we would return an error on line 26 when we try to access it)

File: src/syft/ast/module.py [hunk 2]
```diff
@@ -34,28 +42,20 @@ def add_attr(
...
-        obj_type: Optional[type] = None,
     ) -> Optional[Union[Callable, CallableT]]:
-        if obj_type is not None:
-            if obj_type in self.lookup_cache:
-                path = self.lookup_cache[obj_type]
 
-        if isinstance(path, str):
-            path = path.split(".")
+        _path: List[str] = (
+            path.split(".") if isinstance(path, str) else path if path else []
```
Comment (gmuraru) [inline src/syft/ast/module.py:53]: Q: Also here

File: src/syft/ast/module.py [hunk 3]
```diff
@@ -76,52 +76,68 @@ def add_path(
...
                     return_type_name=return_type_name,
+                    client=self.client,
                 )
                 self.add_attr(
                     attr_name=path[index],
                     attr=klass,
                 )
-            elif isinstance(attr_ref, func_type):
+            elif inspect.isfunction(attr_ref) or inspect.isbuiltin(attr_ref):
+                if path[index] == "T":
```
Comment (gmuraru) [inline src/syft/ast/module.py:101]: MUST: Remove debug :D

File: src/syft/ast/module.py [hunk 4]
```diff
@@ -76,52 +76,68 @@ def add_path(
         if path[index] not in self.attrs:
             attr_ref = getattr(self.ref, path[index])
```
Comment (gmuraru) [inline src/syft/ast/module.py:77]: Q: Maybe create another ```internal method``` for adding all the logic from those ```if```/```else```?

File: src/syft/ast/module.py [hunk 5]
```diff
@@ -76,52 +76,68 @@ def add_path(
...
+                    path.insert(len(path) - 1, "globals")
+                    self.add_attr(
+                        attr_name="globals",
+                        attr=ast.klass.Class(
+                            path_and_name=".".join(path[: index + 1]),
+                            ref=scope,
+                            return_type_name=scope_name,
+                            client=self.client,
+                        ),
+                    )
```
Comment (gmuraru) [inline src/syft/ast/module.py:137]: Q: ?
```
else:
    raise some error
```

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+from ....common.serde.deserialize import _deserialize
+from ....common.uid import UID
+from ....io.address import Address
+from ....store.storeable_object import StorableObject
+from ...abstract.node import AbstractNode
+from .common import ImmediateActionWithoutReply
+from .run_class_method_action import RunClassMethodAction
+
+
+class PropertyActions(Enum):
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:30]: :heart::heart::heart:

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        self, left: Union[Dict[VerifyKey, UID], None], right: Dict[VerifyKey, UID]
+    ) -> Dict[VerifyKey, UID]:
+        return RunClassMethodAction.intersect_keys(left, right)
+
+    def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
+        method = node.lib_ast(self.path)
+        resolved_self = node.store.get_object(key=self._self.id_at_location)
+        result_read_permissions = resolved_self.read_permissions
+
+        resolved_args = list()
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:68]: NITPICK:
```resolved_args = []```

File: src/syft/core/node/common/action/function_or_constructor_action.py [hunk 8]
```diff
@@ -52,9 +52,6 @@ def __init__(
         self.path = path
         self.args = args
         self.kwargs = kwargs
-
-        # TODO: eliminate this explicit parameter and just set the object
-        #  id on the object directly
         self.id_at_location = id_at_location
 
     @staticmethod
```
Comment (gmuraru) [inline src/syft/core/node/common/action/function_or_constructor_action.py:57]: NITPICK: I know you did not touch this code. but why don't we use at line 68:
```
intersection = left.keys() & right.keys()
```

File: src/syft/core/node/common/action/function_or_constructor_action.py [hunk 9]
```diff
@@ -114,23 +110,15 @@ def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
...
-
         # to avoid circular imports
         if lib.python.primitive_factory.isprimitive(value=result):
             # Wrap in a SyPrimitive
             result = lib.python.primitive_factory.PrimitiveFactory.generate_primitive(
                 value=result, id=self.id_at_location
             )
         else:
-            # TODO: overload all methods to incorporate this automatically
             if hasattr(result, "id"):
```
Comment (gmuraru) [inline src/syft/core/node/common/action/function_or_constructor_action.py:120]: NITPICK:
```
elif:
```

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                    if hasattr(result, "_id"):
+                        # set the underlying id
+                        result._id = self.id_at_location
+                    else:
+                        result.id = self.id_at_location
+
+                    assert result.id == self.id_at_location
+                except AttributeError as e:
+                    err = f"Unable to set id on result {type(result)}. {e}"
+                    raise Exception(err)
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:119]: Q: Can't we simply raise again the exception (```raise e```)? (or do we want it to be rewritten?)

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+        if not isinstance(result, StorableObject):
+            result = StorableObject(
+                id=self.id_at_location,
+                data=result,
+                read_permissions=result_read_permissions,
+            )
+
+        node.store[self.id_at_location] = result
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:128]: Q: What happens if it is not a primitive? or we always expect to be a primitive (in that case we can raise an exception if it not primitive?)

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        :rtype: GetOrSetPropertyAction_PB
+        .. note::
+            This method is purely an internal method. Please use object.serialize() or one of
+            the other public serialization methods if you wish to serialize an
+            object.
+        """
+        return GetOrSetPropertyAction_PB(
+            path=self.path,
+            id_at_location=self.id_at_location.serialize(),
+            args=list(map(lambda x: x.serialize(), self.args)),
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:146]: NITPICK: ```[x.serialize() for x in self.args]```

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+            if you wish to deserialize an object.
+        """
+
+        return GetOrSetPropertyAction(
+            path=proto.path,
+            id_at_location=_deserialize(blob=proto.id_at_location),
+            address=_deserialize(blob=proto.address),
+            _self=_deserialize(blob=proto._self),
+            msg_id=_deserialize(blob=proto.msg_id),
+            args=tuple(map(lambda x: _deserialize(blob=x), proto.args)),
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:174]: NITPICK: ```(_deserialize(blob=x) for x in proto.args)```

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+            _self=self._self.serialize(),
+            msg_id=self.id.serialize(),
+            action=self.action.value,
+        )
+
+    @staticmethod
+    def _proto2object(
+        proto: GetOrSetPropertyAction_PB,
+    ) -> "GetOrSetPropertyAction":
+        """Creates a ObjectWithID from a protobuf
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:158]: Q: Creates a ```GetOrSetPropertyAction``` object?

File: src/syft/core/node/common/action/get_or_set_property_action.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+        self.path = path
+        self.id_at_location = id_at_location
+        self.set_arg = set_arg
+        self._self = _self
+        self.action = action
+        self.args = args
+        self.kwargs = kwargs
+
+    def intersect_keys(
+        self, left: Union[Dict[VerifyKey, UID], None], right: Dict[VerifyKey, UID]
```
Comment (gmuraru) [inline src/syft/core/node/common/action/get_or_set_property_action.py:59]: NITPICK: Why not ```Optional[Dict[VerifyKey, UID]```? (I think it is the same with ```Union[..., None]```)

File: src/syft/lib/misc/scope.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
+# stdlib
+import sys
+from types import ModuleType
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import Tuple
+import inspect
+scopes_cache = {}
```
Comment (gmuraru) [inline src/syft/lib/misc/scope.py:9]: NITPICK: Some empty lines after imports

File: src/syft/lib/misc/scope.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+def generate_func(module: ModuleType, method_name: str) -> Callable:
+    def proxy_func(*args: Any, **kwargs: Any) -> Any:
+        object = getattr(module, method_name)
+        return object(*args, **kwargs)
+    return proxy_func
+
+
+class Scope(type):
+    def __new__(cls, name: str, bases: tuple, dct: Dict[str, Any]) -> Any:
```
Comment (gmuraru) [inline src/syft/lib/misc/scope.py:40]: Q: ```bases``` can't be type-hinted with ```Tuple[Something...]```?

**Review by gmuraru — COMMENTED** [review]
With some delay :( (I am sorry!)

**Review by madhavajay — APPROVED** [review]
Wow what a beast! Great job. I have made some minor tweaks and it looks ready for merge. 🎉

**Pull request merged**